### ISSUE TITLE
Enforce type name == constructor name for single-constructor record types.

### ISSUE THREAD

**Issue by associahedron** [digital-asset/daml#4718]
## The Problem

When converting a record type to LF:

```
data Foo = Bar { ... }
```

The type is turned into an LF record type, as one would expect. This causes the constructor name `Bar` to disappear, since there is no place to store it in LF.

This wouldn't normally be a problem, except that in data-dependencies, when we try to go back from LF to DAML, there's no way to tell the constructor name from the available data. Right now, data-dependencies will always use the type name, which is correct more often than not.

In the case above, where the type name isn't the same as the constructor name, data-dependencies will generate the definition:

```
data Foo = Foo { .. }
```

This normally doesn't cause problems, but see below.

## Impact

This has actually caused problems in a large DAML codebase, where the generated constructor's name clashes with a different constructor name:

```
data Foo = Bar { ... }
data X = Foo | Baz
```

In DAML, the `X` enum type has a constructor `Foo`, and the `Foo` record type has a constructor `Bar`. When we go to LF, the `X` type remembers its constructor names, but the `Foo` type does not. When we turn the LF back into DAML for data-dependencies, it becomes

```
data Foo = Foo { ... }
data X = Foo | Baz
```

which causes a compiler error (the constructor `Foo` is defined twice).

Fortunately for this real world case, the record type `Foo` has no fields, and everything works once we change its definition to that of an enum type `data Foo = Bar`, since enums remember their constructor names. The definition `data Foo = Bar {}` was actually left over from when we couldn't support single constructor enums in DAML, so this fix is closer to the original intent behind the code. The fix is not generalizable, however, and different situations would require different approaches.

## Minimal Example

```
data A = B
data B = C {}
```

## Complications

Reconstructing the original constructor name in data-dependencies from the available contextual information isn't really feasible. It introduces a lot of brittle complexity for a rare edge case. We could add meta information like this into LF, but again, that's a lot of added complexity.

## Proposed Solution

The solution we (@cocreature and I) propose is to enforce that the record type name matches the constructor name. We can enforce this syntactically by adding a check in the daml-preprocessor. This eliminates a source of problems and matches the most common use case, where the type and constructor names already match. 

It is a breaking change, however. This could be mitigated by the use of a warning for some time, which later gets upgraded to an error. But this should be weighed against the fact that any DAR or DALF that has this name conflict cannot be used as a data-dependency, so it cannot easily be upgraded. I would favor enforcing this change sooner, rather than later, to prevent problems from propagating any further.

## Possible Interim Solution

A possible interim solution is to convert `data A = B { ... }` into a variant type with a synthetic record constructor, instead of a simple record type. This means it gets converted into a separate variant and record type in LF, as we already do with record constructors for variant types:

```
variant A = A.B A.B
record A.B = { ... }
```

This is something we already have the machinery to do in LFConversion, and we already have to handle this correctly in data-dependencies.

The biggest downside is for users of the ledger protocol and codegens. Users might expect a record type for `A` but would get a variant instead. In this respect, the first solution is much cleaner, and I think we should only consider this second approach as an interim solution, to paper over the breaking change from the first while preventing the generation of DALFs and DARs that are un-upgradeable.

But this is also a breaking change, for users of the ledger protocol and codegens. We could make this change and simultaneously add a warning that "Record type with constructor name different from type name (e.g. `data Foo = Bar { ... }`) is not supported, so it will be treated as a variant type. In the future this will not be permitted."

## Request for Feedback

Thoughts? cc @cocreature @hurryabit @bame-da @gerolf-da

**Comment by cocreature** [comment]
@bame-da approves under the condition that we add a warning.

**Comment by hurryabit** [comment]
I' fine with adding the restriction as well. I'd prefer the deprecation cycle to have ended before we release SDK 1.0 if that's feasible.

**Comment by associahedron** [comment]
#4892 introduces the warning. We can upgrade it to an error after some number of releases.

**Comment by cocreature** [comment]
Fixed by #5266
